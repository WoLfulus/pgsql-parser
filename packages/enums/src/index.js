import toInt from './enum.2int.json';
import toStr from './enum.2str.json';
import nodes from './nodes.json';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
