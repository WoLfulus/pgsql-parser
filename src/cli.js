#!/usr/bin/env node
import { resolve, join } from 'path';
import { readFileSync } from 'fs';
import { parse } from './index';
import { cleanTreeWithStmt } from './utils';
const args = process.argv.slice(2);
if (args.length !== 1) {
  console.warn('Usage: pgsql-parser <sqlfile>');
  process.exit(1);
}

const content = readFileSync(resolve(join(process.cwd(), args[0])), 'utf-8');
const query = parse(content);
process.stdout.write(JSON.stringify(cleanTreeWithStmt(query), null, 2));
