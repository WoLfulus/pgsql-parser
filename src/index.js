import Deparser from './deparser';
const parse = require('@pgql/parse').parseQuerySync;
const parseFunction = require('@pgql/parse').parsePlPgSQLSync;
const deparse = Deparser.deparse;
const enums = require('./libpg_enums');
export { parse, deparse, Deparser, enums, parseFunction };
