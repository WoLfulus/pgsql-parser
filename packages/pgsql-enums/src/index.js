import enums from './libpg_enums';
export { enums };
export * from './types';
