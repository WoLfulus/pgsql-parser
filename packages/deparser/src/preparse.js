/* eslint-disable no-restricted-syntax */

import { getEnum, nodes, toStr, toInt } from 'pgsql-enums';

const wrapRaw = (tree) => {
  return tree.stmts.map(({ stmt, stmt_len }) => {
    return {
      RawStmt: {
        stmt,
        stmt_len
      }
    };
  });
};

export const preparse = (tree) => {
  const stmts = wrapRaw(tree);
  return stmts.map((stmt) => transform2(stmt));
  //   return stmts.map((stmt) => stmt);
};

export const transform = (obj, context) => {
  let copy = null;
  // Handle null or undefined
  if (obj == null) {
    return obj;
  }

  // Handle the 3 simple types
  if (typeof obj !== 'object') {
    // if (context.enum) {
    //   return toInt[context.enum][obj];
    //   return obj;
    // } else {
    return obj;
    // }
  }

  // Handle Date
  if (obj instanceof Date) {
    copy = new Date();
    copy.setTime(obj.getTime());
    return copy;
  }

  // Handle Array
  if (obj instanceof Array) {
    copy = [];
    for (let i = 0, len = obj.length; i < len; i++) {
      copy[i] = transform(obj[i], context);
    }
    return copy;
  }

  // Handle Object
  if (obj instanceof Object || typeof obj === 'object') {
    copy = {};

    for (const attr in obj) {
      // IF IS AST NODE
      if (nodes.hasOwnProperty(attr)) {
        context = { node: attr };
        copy[attr] = transform(obj[attr], context);
      } else {
        // MUST BE A FIELD
        if (context && context.node) {
          const attrSchema = nodes[context.node][attr];

          if (attrSchema?.enum) {
            copy[attr] = transform(obj[attr], {
              enum: attrSchema.type
            });
          } else if (nodes.hasOwnProperty(attrSchema?.type)) {
            if (attrSchema.type === 'Expr') {
              copy[attr] = transform(obj[attr], {
                node: attrSchema.type
              });
            } else {
              copy[attr] = {
                [attrSchema.type]: transform(obj[attr], {
                  node: attrSchema.type
                })
              };
            }
          } else {
            copy[attr] = transform(obj[attr], context);
          }
        } else {
          copy[attr] = transform(obj[attr], context);
        }
      }
    }
    return copy;
  }

  throw new Error("Unable to copy obj! Its type isn't supported.");
};

export const transform2 = (obj, context) => {
  let copy = null;
  // Handle null or undefined
  if (obj == null) {
    return obj;
  }

  // Handle the 3 simple types
  if (typeof obj !== 'object') {
    return obj;
  }

  // Handle Date
  if (obj instanceof Date) {
    copy = new Date();
    copy.setTime(obj.getTime());
    return copy;
  }

  // Handle Array
  if (obj instanceof Array) {
    copy = [];
    for (let i = 0, len = obj.length; i < len; i++) {
      copy[i] = transform2(obj[i], context);
    }
    return copy;
  }

  // Handle Object
  if (obj instanceof Object || typeof obj === 'object') {
    copy = {};

    for (const attr in obj) {
      if (attr === 'List') {
        copy = [];
        for (let i = 0, len = obj[attr].items.length; i < len; i++) {
          copy[i] = transform2(obj[attr].items[i], context);
        }
      } else {
        copy[attr] = transform2(obj[attr], context);
      }
    }
    return copy;
  }

  throw new Error("Unable to copy obj! Its type isn't supported.");
};
