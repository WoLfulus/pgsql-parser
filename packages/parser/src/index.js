import { Deparser, deparse } from 'pgsql-deparser';
import { enums } from 'pgsql-enums';
const parse = require('libpg-query').parseQuerySync;
const parseFunction = require('libpg-query').parsePlPgSQLSync;
export { parse, deparse, Deparser, enums, parseFunction };
