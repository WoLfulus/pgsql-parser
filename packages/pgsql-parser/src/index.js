import { Deparser, deparse } from 'pgsql-deparser';
import { enums } from 'pgsql-enums';
const parse = require('@pgql/parse').parseQuerySync;
const parseFunction = require('@pgql/parse').parsePlPgSQLSync;
export { parse, deparse, Deparser, enums, parseFunction };
