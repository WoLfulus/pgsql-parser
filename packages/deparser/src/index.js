import Deparser from './deparser';
import { preparse } from './preparse';
const deparse = Deparser.deparse;
export { deparse, Deparser, preparse };
