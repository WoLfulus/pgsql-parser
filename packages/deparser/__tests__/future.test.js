import { deparse } from '../src';
import { cleanTree, cleanLines } from '../src/utils';
import { readFileSync } from 'fs';
import { sync as glob } from 'glob';

const FIXTURE_DIR = `${__dirname}/../../../__future__`;

export const check = (file) => {
  const testsql = glob(`${FIXTURE_DIR}/${file}*.sql`).map((f) =>
    readFileSync(f).toString()
  )[0];
  const testjson = glob(`${FIXTURE_DIR}/${file}*.json`).map((f) =>
    JSON.parse(readFileSync(f, 'utf-8'))
  )[0];

  console.log(JSON.stringify(testjson, null, 2));
  const sql = deparse(testjson);
  console.log(sql);
  expect(sql).toMatchSnapshot();
};

it('include-index', () => {
  check('include-index');
});


