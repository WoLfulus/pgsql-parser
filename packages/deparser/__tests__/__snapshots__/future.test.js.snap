// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generated 1`] = `""`;

exports[`include-index 1`] = `"CREATE UNIQUE INDEX new_unique_idx ON new_example ( a, b ) INCLUDE ( c );"`;
